src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -5846,7 +5870,13 @@ private BoundExpression MakeBadExpressionForObjectCreation(SyntaxNode node, Type
                         hasErrors |= !CheckNestedObjectInitializerPropertySymbol(property, leftSyntax, diagnostics, hasErrors, ref resultKind);
                     }
 
-                    return hasErrors ? boundMember : CheckValue(boundMember, valueKind, diagnostics);
+                    return hasErrors ?
+                               boundMember :
+                               CheckValue(boundMember, valueKind, diagnostics
+#if DEBUG
+                                          , dynamificationOfAssignmentResultIsHandled: boundMember is not BoundIndexerAccess
```

Comment:
For the various `dynamificationOfAssignmentResultIsHandled: boundMember is not BoundIndexerAccess` instances, I think a direct assertion `Debug.Assert(boundMember is not BoundIndexerAccess)` or a comment would be helpful to clarify the intent.